spring-integration-ip/src/test/java/org/springframework/integration/ip/tcp/connection/FailoverClientConnectionFactoryTests.java

```diff
@@ -193,10 +201,52 @@ public void testFailoverAllDead() throws Exception {
 		FailoverClientConnectionFactory failoverFactory = new FailoverClientConnectionFactory(factories);
 		failoverFactory.start();
 		GenericMessage<String> message = new GenericMessage<String>("foo");
-		failoverFactory.getConnection().send(message);
+		assertThatExceptionOfType(UncheckedIOException.class).isThrownBy(() ->
+			failoverFactory.getConnection().send(message));
 		Mockito.verify(conn2).send(message);
 	}
 
+	@LogLevels(level = "DEBUG", classes =
+			{ FailoverClientConnectionFactory.class, TcpNetClientConnectionFactory.class, TcpNetConnection.class },
+			categories =
+			"org.springframework.integration.ip.tcp.connection.FailoverClientConnectionFactory$FailoverTcpConnection")
+	@Test
+	void failoverAllDeadAfterSuccess() throws Exception {
+		ServerSocket ss1 = ServerSocketFactory.getDefault().createServerSocket(0);
+		ServerSocket ss2 = ServerSocketFactory.getDefault().createServerSocket(0);
+		int port2 = ss2.getLocalPort();
+		ss2.close();
+		ThreadPoolTaskExecutor exec = new ThreadPoolTaskExecutor();
+		exec.initialize();
+		Future<Boolean> done = exec.submit(() -> {
+			Socket accepted = ss1.accept();
+			BufferedReader br = new BufferedReader(new InputStreamReader(accepted.getInputStream()));
+			br.readLine();
+			accepted.getOutputStream().write("ok\r\n".getBytes());
+			accepted.close();
+			ss1.close();
+			return true;
+		});
+		TcpNetClientConnectionFactory cf1 = new TcpNetClientConnectionFactory("localhost", ss1.getLocalPort());
+		TcpNetClientConnectionFactory cf2 = new TcpNetClientConnectionFactory("localhost", port2);
+		CountDownLatch latch = new CountDownLatch(2);
+		cf1.setApplicationEventPublisher(event -> {
+			latch.countDown();
+		});
+		cf2.setApplicationEventPublisher(event -> { });
+		FailoverClientConnectionFactory fccf = new FailoverClientConnectionFactory(List.of(cf1, cf2));
+		fccf.registerListener(msf -> {
+			latch.countDown();
+			return false;
+		});
+		fccf.start();
+		fccf.getConnection().send(new GenericMessage<>("test"));
+		assertThat(done.get(10, TimeUnit.SECONDS)).isTrue();
```

Comment:
Do we need this extra barrier to wait?
I think that `latch` on the next line is enough for this test.